### ISSUE TITLE
osmdroid overlay server requests burst up

### ISSUE THREAD

**Issue by jeiea** [osmdroid/osmdroid#1167]
## Issue Type

- [x] Question

## Description and/or steps/code to reproduce the problem

I'm using a MapView with regular invalidation and this is the web traffic record.
![ccap](https://user-images.githubusercontent.com/4188984/46726699-d935d500-ccb9-11e8-851c-cc072417ef71.PNG)
It seems a cache does no work. I thought the server's time and the expires header value could be a problem.
![cc](https://user-images.githubusercontent.com/4188984/46726757-fcf91b00-ccb9-11e8-98d5-7e595d35d502.PNG)
Is this problem affected only to me?

## Environment

Emulator, Android API 27

### If it's a bug, version(s) of android this affects:

Not sure

### Version of osmdroid the issue relates to:

6.0.2

**Comment by monsieurtanuki** [comment]
@jeiea What do you mean by "regular invalidation"? Something like `View.invalidate`?
Do you use the `CacheManager` (bulk download), or do you just pan the map (or even just do nothing)?
I'm not familiar with your web traffic record: what's the typical time gap between 2 records? Less than a second?
What tile providers do you use - can you share your code specifically on that topic?
Given the header values, the tiles will ask for a refresh every 2 days.

**Comment by jeiea** [comment]
@monsieurtanuki First here is the sample code.
```kotlin
package com.github.jeiea.justosmdroid

import android.os.Bundle
import android.preference.PreferenceManager
import android.support.v7.app.AppCompatActivity
import org.osmdroid.config.Configuration
import org.osmdroid.tileprovider.tilesource.TileSourceFactory
import org.osmdroid.views.MapView


class MainActivity : AppCompatActivity() {

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    val ctx = applicationContext
    val default = PreferenceManager.getDefaultSharedPreferences(ctx)
    Configuration.getInstance().load(ctx, default)

    setContentView(R.layout.activity_main)
    val map = findViewById<MapView>(R.id.map)
    map.setTileSource(TileSourceFactory.BASE_OVERLAY_NL)
    map.controller.setZoom(3.0)
  }
}
```

activity_main.xml contains just a simple `MapView`. It seems invalidation has no relation.

I used fiddler2 software, and unfortunately it doesn't provide time column feature. But the time gap is about 1~2 seconds.

I tried modifying response header value, and this situation was fixed.

**Comment by monsieurtanuki** [comment]
@jeiea I've just run some tests using `SampleWithTilesOverlay` on `TileSourceFactory.BASE_OVERLAY_NL` with zoom level 7 (today is 2018.10.25)

Checking with a random http header website, this is what I've found for `http://overlay.openstreetmap.nl/basemap/7/65/42.png`:

> HTTP/1.0 200 OK =>
Connection => close
Date => Wed, 17 Oct 2018 17:47:00 GMT
Server => Cherokee/1.2.104 (UNIX)
Content-Type => image/png
Content-Length => 31075
Expires => Fri, 19 Oct 2018 17:47:00 GMT
Cache-Control => max-age=172800, public

In `TileDownloader.downloadTile`, regarding tile expiration, among http headers we only check `Expires`. And the value (2010.10.19 17:47) is expired, therefore in your case it's normal that the tiles are downloaded again and again.

But from what I read about [Expires http header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Expires), we should use `max-age` instead when it's available:
> If there is a Cache-Control header with the "max-age" or "s-maxage" directive in the response, the Expires header is ignored.

In this example `max-age` is `172800`, therefore 48 hours: the tiles should expire 2 days later.

About [Cache-Control http header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control):
> max-age=<seconds>
    Specifies the maximum amount of time a resource will be considered fresh. Contrary to Expires, this directive is relative to the time of the request.

I guess `TileDownloader.downloadTile` should be amended in order to take into account the `Cache-Control: max-age` http header first, then `Expires`.

As a quick fix in your case, there are cheats in order to extend the expiration delay, like `Configuration.getInstance().setExpirationExtendedDuration()` (number of milliseconds to add to the `Expires` http header in order to compute the tile expiration date).

**Comment by jeiea** [comment]
@monsieurtanuki Thanks for the workaround. I applied your solution.

**Comment by monsieurtanuki** [comment]
Working on the fix.

**Comment by monsieurtanuki** [comment]
Fixed by #1173.
